### PR TITLE
gateway: balance, address, deposit & withdraw CLI commands

### PR DESCRIPTION
**Pull request by justinmoon**
This PR has a lot of copy-pasted code which should be removed by https://github.com/fedimint/minimint/issues/236

Also, some of the message passing code is a little boilerplatey.

### CODE REVIEW

File: client/client-lib/src/clients/gateway.rs [hunk 1]
```diff
@@ -361,6 +374,102 @@ impl GatewayClient {
         self.context.db.apply_batch(batch).expect("DB error");
         Ok(())
     }
+
+    // FIXME: copy-pasted from UserClient
```
Comment (jkitman) [inline client/client-lib/src/clients/gateway.rs:378]: What if we just gave the `GatewayClient` an instance of `UserClient`?  That's what we do in the integration tests.

Comment (elsirion) [reply client/client-lib/src/clients/gateway.rs:378]: @justinmoon is fixing this rn (in another PR afaik). The idea is to have three type of configs:

```rust
struct ClientConfig {
…
}

struct UserClientConfig {
    common: ClientConfig,
    …
}

struct GatewayClientConfig {
    common: ClientConfig,
    …
}

impl AsRef<ClientConfig> for UserClientConfig {…}
impl AsRef<ClientConfig> for GatewayClientConfig {…}
```

With that we can have a client generic over a config with three implementations, a common and two specialized ones:

```rust
struct Client<T> {
    config: T,
    …
}

impl<T: AsRef<ClientConfig>> for Client<T> {
    // Common functionality
}

impl Client<UserClientConfig> {
    // User specific functionality
}

impl Client<GatewayClientConfig> {
    // Gateway specific functionality
}
```

File: ln-gateway/src/bin/ln_gateway.rs [hunk 2]
```diff
@@ -115,21 +135,91 @@ async fn htlc_accepted_handler(
...
+    let gw_sender = { plugin.state().lock().await.clone() };
+    gw_sender
+        .send(GatewayRequest::Balance(sender))
+        .await
+        .expect("failed to send over channel");
+    let amount = receiver.await.unwrap()?;
+    Ok(json!({ "balance_msat": amount.milli_sat }))
+}
+
+async fn address(
```
Comment (elsirion) [inline ln-gateway/src/bin/ln_gateway.rs:167]: Should be named a bit more informatively, like `new_deposit_address`.

File: ln-gateway/src/bin/ln_gateway.rs [hunk 3]
```diff
@@ -115,21 +135,91 @@ async fn htlc_accepted_handler(
...
+) -> Result<serde_json::Value, Error> {
+    let deposit: Deposit = serde_json::from_value(value)?;
+    let (sender, receiver) = oneshot::channel::<Result<TransactionId, LnGatewayError>>();
+    let gw_sender = { plugin.state().lock().await.clone() };
+    gw_sender
+        .send(GatewayRequest::Deposit(deposit.0, deposit.1, sender))
+        .await
+        .expect("failed to send over channel");
+    let txid = receiver.await.unwrap()?;
+    Ok(json!({ "fedimint_txid": txid.to_string() }))
```
Comment (elsirion) [inline ln-gateway/src/bin/ln_gateway.rs:199]: I wonder if I can build a nice higher order function to deduplicate that code :smile:

File: client/client-lib/src/clients/gateway.rs (hunk 1 above)
Comment (justinmoon) [reply client/client-lib/src/clients/gateway.rs]: Fixed by #260

File: scripts/gw-pegin.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env bash
+# Calls the CLI to peg gateway client into the federation
+
+set -euxo pipefail
+
+# Let's define some shortcuts for bitcoind and the mint client
+POLL_INTERVAL=1
+PEG_IN_AMOUNT=${PEG_IN_AMOUNT:-$1}
```
[the remark this answers is not in the document]
Comment (justinmoon) [reply scripts/gw-pegin.sh]: Done

File: scripts/lib.sh [hunk 5]
```diff
@@ -13,4 +13,15 @@ function open_channel() {
     until $FM_LN1 -k fundchannel id=$FM_LN2_PUB_KEY amount=0.1btc push_msat=5000000000; do sleep $POLL_INTERVAL; done
     mine_blocks 10
     until [[ $($FM_LN1 listpeers | jq -r ".peers[] | select(.id == \"$FM_LN2_PUB_KEY\") | .channels[0].state") = "CHANNELD_NORMAL" ]]; do sleep $POLL_INTERVAL; done
+}
+
+function await_block_sync() {
+  EXPECTED_BLOCK_HEIGHT="$(( $($FM_BTC_CLIENT getblockchaininfo | jq -r '.blocks') - $CONFIRMATION_TIME ))"
```
Comment (justinmoon) [inline scripts/lib.sh:19]: ci failing because `CONFIRMATION_TIME` not defined here ... will look tomorrow

**Comment by justinmoon** [comment]
CI passing now

**Review by jkitman — APPROVED** [review]
LGTM!

**Review by elsirion — APPROVED** [review]

**Pull request merged**